utils/pdftops.1
.\" Copyright 1996-2011 Glyph & Cog, LLC
.TH pdftops 1 "15 August 2011"
.SH NAME
pdftops \- Portable Document Format (PDF) to PostScript converter
(version 3.03)
.SH SYNOPSIS
.B pdftops
[options]
.RI <PDF-file>
.RI [<PS-file>]
.SH DESCRIPTION
.B Pdftops
converts Portable Document Format (PDF) files to PostScript so they
can be printed.
.PP
Pdftops reads the PDF file,
.IR PDF-file ,
and writes a PostScript file,
.IR PS-file .
If
.I PS-file
is not specified, pdftops converts
.I file.pdf
to
.I file.ps
(or
.I file.eps
with the \-eps option).  If
.I PS-file
is \'-', the PostScript is sent to stdout.  If
.I PDF-file
is \'-', Pdftops reads the PDF file from stdin.
.SH OPTIONS
.TP
.BI \-f " number"
Specifies the first page to print.
.TP
.BI \-l " number"
Specifies the last page to print.
.TP
.B \-level1
Generate Level 1 PostScript.  The resulting PostScript files will be
significantly larger (if they contain images), but will print on Level
1 printers.  This also converts all images to black and white.  No
more than one of the PostScript level options (\-level1, \-level1sep,
\-level2, \-level2sep, \-level3, \-level3sep) may be given.
.TP
.B \-level1sep
Generate Level 1 separable PostScript.  All colors are converted to
CMYK.  Images are written with separate stream data for the four
components.
.TP
.B \-level2
Generate Level 2 PostScript.  Level 2 supports color images and image
compression.  This is the default setting.
.TP
.B \-level2sep
Generate Level 2 separable PostScript.  All colors are converted to
CMYK.  The PostScript separation convention operators are used to
handle custom (spot) colors.
.TP
.B \-level3
Generate Level 3 PostScript.  This enables all Level 2 features plus
CID font embedding.
.TP
.B \-level3sep
Generate Level 3 separable PostScript.  The separation handling is the
same as for \-level2sep.
.TP
.B \-eps
Generate an Encapsulated PostScript (EPS) file.  An EPS file contains
a single image, so if you use this option with a multi-page PDF file,
you must use \-f and \-l to specify a single page.  No more than one of
the mode options (\-eps, \-form) may be given.
.TP
.B \-form
Generate a PostScript form which can be imported by software that
understands forms.  A form contains a single page, so if you use this
option with a multi-page PDF file, you must use \-f and \-l to specify a
single page.  The \-level1 option cannot be used with \-form.  No more
than one of the mode options (\-eps, \-form) may be
given.
.TP
.B \-opi
Generate OPI comments for all images and forms which have OPI
information.  (This option is only available if pdftops was compiled
with OPI support.)
.TP
.B \-binary
Write binary data in Level 1 PostScript.  By default, pdftops writes
hex-encoded data in Level 1 PostScript.  Binary data is non-standard
in Level 1 PostScript but reduces the file size and can be useful
when Level 1 PostScript is required only for its restricted use
of PostScript operators.
.TP
.BI \-r " number"
Set the resolution in DPI when pdftops rasterizes images with
transparencies or, for Level 1 PostScript, when pdftops
rasterizes images with color masks.
By default, pdftops rasterizes images to 300 DPI.
.TP
.B \-noembt1
By default, any Type 1 fonts which are embedded in the PDF file are
copied into the PostScript file.  This option causes pdftops to
substitute base fonts instead.  Embedded fonts make PostScript files
larger, but may be necessary for readable output.
.TP
.B \-noembtt
By default, any TrueType fonts which are embedded in the PDF file are
copied into the PostScript file.  This option causes pdftops to
substitute base fonts instead.  Embedded fonts make PostScript files
larger, but may be necessary for readable output.  Also, some
PostScript interpreters do not have TrueType rasterizers.
.TP
.B \-noembcidps
By default, any CID PostScript fonts which are embedded in the PDF
file are copied into the PostScript file.  This option disables that
embedding.  No attempt is made to substitute for non-embedded CID
PostScript fonts.
.TP
.B \-noembcidtt
By default, any CID TrueType fonts which are embedded in the PDF file
are copied into the PostScript file.  This option disables that
embedding.  No attempt is made to substitute for non-embedded CID
TrueType fonts.
.TP
.B \-passfonts
By default, references to non-embedded 8-bit fonts in the PDF file are
substituted with the closest "Helvetica", "Times-Roman", or "Courier" font.
This option passes references to non-embedded fonts
through to the PostScript file.
.TP
.BI \-aaRaster " yes | no"
Enable or disable raster anti-aliasing.  This defaults to "no".
pdftops may need to rasterize transparencies and pattern image masks in the PDF.
If the PostScript will be printed, leave \-aaRaster disabled and set \-r to the resolution of the printer.
If the PostScript will be viewed, enabling \-aaRaster may make rasterized text easier to read.
.TP
.BI \-rasterize " always | never | whenneeded"
By default, pdftops rasterizes pages as needed, for example, if they contain transparencies.
To force rasterization, set \-rasterize to "always". Use this to eliminate fonts.
To prevent rasterization, set \-rasterize to "never". This may produce files that display incorrectly.
.TP
.BI \-processcolorformat " MONO8 | CMYK8 | RGB8"
Sets the process color format as it is used during rasterization and transparency reduction.
The default depends on the other settings: For \-level1 the default is MONO8, for \-level{1,2,3}sep
or \-overprint the default is CMYK8, and in all other cases RGB8 is the default. If \-processcolorprofile
is given then \-processcolorformat is inferred from the specified ICC profile.
.TP
.BI \-processcolorprofile " filename"
Sets the ICC profile that is assumed during rasterization and transparency reduction.
.TP
.B \-optimizecolorspace
By default, bitmap images in the PDF pass through to the output PostScript
in their original color space, which produces predictable results.
This option converts RGB and CMYK images into Gray images
if every pixel of the image has equal components.
This can fix problems when doing color separations of PDFs
that contain embedded black and white images encoded as RGB.
.TP
.B \-preload
preload images and forms
.TP
.BI \-paper " size"
Set the paper size to one of "letter", "legal", "A4", or "A3".  This
can also be set to "match", which will set the paper size of each page to match the
size specified in the PDF file. If none the \-paper, \-paperw, or \-paperh
options are specified the default is to match the paper size.
.TP
.BI \-paperw " size"
Set the paper width, in points.
.TP
.BI \-paperh " size"
Set the paper height, in points.
.TP
.B \-origpagesizes
This option is the same as "\-paper match".
.TP
.B \-nocrop
By default, output is cropped to the CropBox specified in the PDF
file.  This option disables cropping.
.TP
.B \-expand
Expand PDF pages smaller than the paper to fill the paper.  By
default, these pages are not scaled.
.TP
.B \-noshrink
Don't scale PDF pages which are larger than the paper.  By default,
pages larger than the paper are shrunk to fit.
.TP
.B \-nocenter
By default, PDF pages smaller than the paper (after any scaling) are
centered on the paper.  This option causes them to be aligned to the
lower-left corner of the paper instead.
.TP
.B \-duplex
Set the Duplex pagedevice entry in the PostScript file.  This tells
duplex-capable printers to enable duplexing.
.TP
.BI \-opw " password"
Specify the owner password for the PDF file.  Providing this will
bypass all security restrictions.
.TP
.BI \-upw " password"
Specify the user password for the PDF file.
.TP
.B \-overprint
Enable overprinting.
.TP
.B \-q
Don't print any messages or errors.
.TP
.B \-v
Print copyright and version information.
.TP
.B \-h
Print usage information.
.RB ( \-help
and
.B \-\-help
are equivalent.)
.SH EXIT CODES
The Xpdf tools use the following exit codes:
.TP
0
No error.
.TP
1
Error opening a PDF file.
.TP
2
Error opening an output file.
.TP
3
Error related to PDF permissions.
.TP
99
Other error.
.SH AUTHOR
The pdftops software and documentation are copyright 1996-2011 Glyph &
Cog, LLC.
.SH "SEE ALSO"
.BR pdfdetach (1),
.BR pdffonts (1),
.BR pdfimages (1),
.BR pdfinfo (1),
.BR pdftocairo (1),
.BR pdftohtml (1),
.BR pdftoppm (1),
.BR pdftotext (1)
.BR pdfseparate (1),
.BR pdfsig (1),
.BR pdfunite (1)


NEWS
Release 0.88.0:
        core:
         * Support Widget Annotation Buttons not linked to any Form
         * SplashOuputDev: Use stroking opacity when clipping to a stroke path
         * Handle 1 bit RGB images in ICC colorspace
         * Internal code improvements

        qt5:
         * Add Document::signatures. Returns signatures not attached to any page
         * ArthurOutputDev: Fix font hinting
         * ArthurOutputDev: Set the opacity when filling with axial gradients
         * ArthurOutputDev: Implement the clipToStrokePath method
         * ArthurOutputDev: Use stroking opacity when clipping to a stroke path

        glib:
         * Add poppler_page_get_bounding_box
         * Add poppler_form_field_get_alternate_ui_name
         * Implement rotation for 'flagNoRotate' annots. Issue #767

        cpp:
         * Add non_raw_non_physical layout for page::text()

        utils:
         * pdftohtml: Fix noRoundedCoordinates->noroundcoord in man page
         * pdfsig: Show also signatures that aren't attached to any page

Release 0.87.0:
        core:
         * Fix leak in broken files
         * Internal code improvements

        qt5:
         * Add option to get form choice for export value
         * ArthurOutputDev: Avoid division by zero in updateLineDash. Issue #695

        glib:
         * Internal code improvements

         utils:
         * pdftohtml: Fix memory leak in HtmlOutputDev::getLinkDest

Release 0.86.1:
        core:
         * Fix regression in Browse Link handling
         * Internal code improvements

Release 0.86.0:
        core:
         * Fix link content exfiltration attack
         * Splash: Implement gouraudTriangleShadedFill for some non parametrized shadings. Issue #881
         * Fix case unsensitive search for Old Hungarian, Warang Citi, Medefaidrin and Adlam
         * Internal code improvements

        glib:
         * Automatic handle of page's cropbox on annots. Issue #129
         * Fix memory leak if poppler_document_new_from_file fails
         * Minor speed optimization on poppler_page_get_annot_mapping

         utils:
         * pdfdetach: add 'savefile' option
         * pdftoppm/pdftocairo: Fix more odd/even mismatch

        qt5:
         * Fix loading from iodevice

Release 0.85.0:
        core:
         * Fix case unsensitive search for Deseret and Osage. Issue #853
         * Fix crash in unicodeToAscii7
         * CairoOutputDev: make initialisation thread-safe
         * Fix crash on broken files. Issues #869, #870
         * Internal code improvements

        utils:
         * pdftoppm/pdftocairo: Fix -e/-o printing the wrong pages. Issue #873
         * pdftohtml: Fix issue with the font size sometimes being huge

        qt5:
         * Fix FormField::name encoding
         * Accept UTF-16 uiNames for form fields
         * Fix search for "complex" characters
         * Allow to load document from QIODevice

        glib:
         * make the frontend initialization thread safe.

Release 0.84.0:
        core:
         * Fix crash when converting from Unicode to ASCII-7
         * Splash::scaleImageYdXu: Protect against crash if srcWidth is too big
         * JBIG2Stream: fix potential crash in malformed documents
         * JBIG2Stream: fix leak in reset() if called several times
         * Internal code improvements

        utils:
         * pdfimages: Add error message if first page is larger then number of pages.
         * pdfinfo: Improved paper size recognition
         * pdfsig: Fix exit code when dumping signatures
         * pdftocairo: Error out when even/odd selects 0 pages
         * pdftohtml: Fix memory leak
         * pdftoppm: Add an option to scale before rotate
         * pdftoppm: Add -hide-annotations option
         * pdftoppm: Error out when even/odd selects 0 pages
         * pdftops: Improve -optimizecolorspace

        qt5:
         * Code cleanups

        glib:
         * Fix compiler warrnings

Release 0.83.0:
        core:
         * Improve when a file is recognized as Linearized
         * Improve const-ness of the code
         * Make code a bit more readable/maintanable
         * Fix uninitialized memory uses in broken files

        utils:
         * pdffonts: Make code a bit more readable/maintanable
         * pdftohtml: Make code a bit more readable/maintanable

        qt5:
         * Remove a bunch of unused internal functions
         * trUtf8 -> tr (less warnings)

        build system:
         * make-glib-api-docs: switch to python3

Release 0.82.0:
        core:
         * Fix not being able to open some files. Issue #832
         * Fix crashes in malformed files
         * Fix memory leak on broken files
         * Minor performance improvements
         * Minor code improvements

        glib:
         * Add poppler_document_new_from_bytes
         * PopplerAttachment: Silence deprecation warnings for ctime/mtime

        build system:
         * pdf-inspector: Support builddir != srcdir
         * Install Cairo* headers if Cairo has been found

Release 0.81.0:
        core:
         * Splash: Always enable support for CMYK rasterization
         * CairoOutputDev: Check scaled dimensions for 0. Issue #737
         * BaseCryptStream: Fix potential uninitialized memory read

utils/pdftops.cc
//========================================================================
//
// pdftops.cc
//
// Copyright 1996-2003 Glyph & Cog, LLC
//
// Modified for Debian by Hamish Moffatt, 22 May 2002.
//
//========================================================================

//========================================================================
//
// Modified under the Poppler project - http://poppler.freedesktop.org
//
// All changes made under the Poppler project to this file are licensed
// under GPL version 2 or later
//
// Copyright (C) 2006 Kristian Høgsberg <krh@redhat.com>
// Copyright (C) 2007-2008, 2010, 2015, 2017, 2018 Albert Astals Cid <aacid@kde.org>
// Copyright (C) 2009 Till Kamppeter <till.kamppeter@gmail.com>
// Copyright (C) 2009 Sanjoy Mahajan <sanjoy@mit.edu>
// Copyright (C) 2009, 2011, 2012, 2014-2016, 2020 William Bader <williambader@hotmail.com>
// Copyright (C) 2010 Hib Eris <hib@hiberis.nl>
// Copyright (C) 2012 Thomas Freitag <Thomas.Freitag@alfa.de>
// Copyright (C) 2013 Suzuki Toshiya <mpsuzuki@hiroshima-u.ac.jp>
// Copyright (C) 2014, 2017 Adrian Johnson <ajohnson@redneon.com>
// Copyright (C) 2018 Adam Reichold <adam.reichold@t-online.de>
// Copyright (C) 2019 Oliver Sander <oliver.sander@tu-dresden.de>
//
// To see a description of the changes please see the Changelog file that
// came with your tarball or type make ChangeLog if you are building from git
//
//========================================================================

#include "config.h"
#include <poppler-config.h>
#include <cstdio>
#include <cstdlib>
#include <cstddef>
#include <cstring>
#include "parseargs.h"
#include "goo/GooString.h"
#include "goo/gmem.h"
#include "GlobalParams.h"
#include "Object.h"
#include "Stream.h"
#include "Array.h"
#include "Dict.h"
#include "XRef.h"
#include "Catalog.h"
#include "Page.h"
#include "PDFDoc.h"
#include "PDFDocFactory.h"
#include "PSOutputDev.h"
#include "Error.h"
#include "Win32Console.h"

#ifdef USE_CMS
#    include <lcms2.h>
#endif

static bool setPSPaperSize(char *size, int &psPaperWidth, int &psPaperHeight)
{
    if (!strcmp(size, "match")) {
        psPaperWidth = psPaperHeight = -1;
    } else if (!strcmp(size, "letter")) {
        psPaperWidth = 612;
        psPaperHeight = 792;
    } else if (!strcmp(size, "legal")) {
        psPaperWidth = 612;
        psPaperHeight = 1008;
    } else if (!strcmp(size, "A4")) {
        psPaperWidth = 595;
        psPaperHeight = 842;
    } else if (!strcmp(size, "A3")) {
        psPaperWidth = 842;
        psPaperHeight = 1190;
    } else {
        return false;
    }
    return true;
}

static int firstPage = 1;
static int lastPage = 0;
static bool level1 = false;
static bool level1Sep = false;
static bool level2 = false;
static bool level2Sep = false;
static bool level3 = false;
static bool level3Sep = false;
static bool origPageSizes = false;
static bool doEPS = false;
static bool doForm = false;
#ifdef OPI_SUPPORT
static bool doOPI = false;
#endif
static int splashResolution = 0;
static bool psBinary = false;
static bool noEmbedT1Fonts = false;
static bool noEmbedTTFonts = false;
static bool noEmbedCIDPSFonts = false;
static bool noEmbedCIDTTFonts = false;
static bool fontPassthrough = false;
static bool optimizeColorSpace = false;
static bool passLevel1CustomColor = false;
static char rasterAntialiasStr[16] = "";
static char forceRasterizeStr[16] = "";
static bool preload = false;
static char paperSize[15] = "";
static int paperWidth = -1;
static int paperHeight = -1;
static bool noCrop = false;
static bool expand = false;
static bool noShrink = false;
static bool noCenter = false;
static bool duplex = false;
static char ownerPassword[33] = "\001";
static char userPassword[33] = "\001";
static bool quiet = false;
static bool printVersion = false;
static bool printHelp = false;
static bool overprint = false;
#ifdef HAVE_SPLASH
static GooString processcolorformatname;
static SplashColorMode processcolorformat = splashModeUndefined;
#    ifdef USE_CMS
static GooString processcolorprofilename;
static GfxLCMSProfilePtr processcolorprofile;
#    endif
#endif

static const ArgDesc argDesc[] = { { "-f", argInt, &firstPage, 0, "first page to print" },
                                   { "-l", argInt, &lastPage, 0, "last page to print" },
                                   { "-level1", argFlag, &level1, 0, "generate Level 1 PostScript" },
                                   { "-level1sep", argFlag, &level1Sep, 0, "generate Level 1 separable PostScript" },
                                   { "-level2", argFlag, &level2, 0, "generate Level 2 PostScript" },
                                   { "-level2sep", argFlag, &level2Sep, 0, "generate Level 2 separable PostScript" },
                                   { "-level3", argFlag, &level3, 0, "generate Level 3 PostScript" },
                                   { "-level3sep", argFlag, &level3Sep, 0, "generate Level 3 separable PostScript" },
                                   { "-origpagesizes", argFlag, &origPageSizes, 0, "conserve original page sizes" },
                                   { "-eps", argFlag, &doEPS, 0, "generate Encapsulated PostScript (EPS)" },
                                   { "-form", argFlag, &doForm, 0, "generate a PostScript form" },
#ifdef OPI_SUPPORT
                                   { "-opi", argFlag, &doOPI, 0, "generate OPI comments" },
#endif
                                   { "-r", argInt, &splashResolution, 0, "resolution for rasterization, in DPI (default is 300)" },
                                   { "-binary", argFlag, &psBinary, 0, "write binary data in Level 1 PostScript" },
                                   { "-noembt1", argFlag, &noEmbedT1Fonts, 0, "don't embed Type 1 fonts" },
                                   { "-noembtt", argFlag, &noEmbedTTFonts, 0, "don't embed TrueType fonts" },
                                   { "-noembcidps", argFlag, &noEmbedCIDPSFonts, 0, "don't embed CID PostScript fonts" },
                                   { "-noembcidtt", argFlag, &noEmbedCIDTTFonts, 0, "don't embed CID TrueType fonts" },
                                   { "-passfonts", argFlag, &fontPassthrough, 0, "don't substitute missing fonts" },
                                   { "-aaRaster", argString, rasterAntialiasStr, sizeof(rasterAntialiasStr), "enable anti-aliasing on rasterization: yes, no" },
                                   { "-rasterize", argString, forceRasterizeStr, sizeof(forceRasterizeStr), "control rasterization: always, never, whenneeded" },
#ifdef HAVE_SPLASH
                                   { "-processcolorformat", argGooString, &processcolorformatname, 0, "color format that is used during rasterization and transparency reduction: MONO8, RGB8, CMYK8" },
#    ifdef USE_CMS
                                   { "-processcolorprofile", argGooString, &processcolorprofilename, 0, "ICC color profile to use as the process color profile during rasterization and transparency reduction" },
#    endif
#endif
                                   { "-optimizecolorspace", argFlag, &optimizeColorSpace, 0, "convert gray RGB images to gray color space" },
                                   { "-passlevel1customcolor", argFlag, &passLevel1CustomColor, 0, "pass custom color in level1sep" },
                                   { "-preload", argFlag, &preload, 0, "preload images and forms" },
                                   { "-paper", argString, paperSize, sizeof(paperSize), "paper size (letter, legal, A4, A3, match)" },
                                   { "-paperw", argInt, &paperWidth, 0, "paper width, in points" },
                                   { "-paperh", argInt, &paperHeight, 0, "paper height, in points" },
                                   { "-nocrop", argFlag, &noCrop, 0, "don't crop pages to CropBox" },
                                   { "-expand", argFlag, &expand, 0, "expand pages smaller than the paper size" },
                                   { "-noshrink", argFlag, &noShrink, 0, "don't shrink pages larger than the paper size" },
                                   { "-nocenter", argFlag, &noCenter, 0, "don't center pages smaller than the paper size" },
                                   { "-duplex", argFlag, &duplex, 0, "enable duplex printing" },
                                   { "-opw", argString, ownerPassword, sizeof(ownerPassword), "owner password (for encrypted files)" },
                                   { "-upw", argString, userPassword, sizeof(userPassword), "user password (for encrypted files)" },
                                   { "-overprint", argFlag, &overprint, 0, "enable overprint emulation during rasterization" },
                                   { "-q", argFlag, &quiet, 0, "don't print any messages or errors" },
                                   { "-v", argFlag, &printVersion, 0, "print copyright and version info" },
                                   { "-h", argFlag, &printHelp, 0, "print usage information" },
                                   { "-help", argFlag, &printHelp, 0, "print usage information" },
                                   { "--help", argFlag, &printHelp, 0, "print usage information" },
                                   { "-?", argFlag, &printHelp, 0, "print usage information" },
                                   {} };

int main(int argc, char *argv[])
{
    PDFDoc *doc;
    GooString *fileName;
    GooString *psFileName;
    PSLevel level;
    PSOutMode mode;
    GooString *ownerPW, *userPW;
    PSOutputDev *psOut;
    bool ok;
    int exitCode;
    bool rasterAntialias = false;
    std::vector<int> pages;
#ifdef USE_CMS
    cmsColorSpaceSignature displayprofilecolorspace;
#endif

    Win32Console win32Console(&argc, &argv);
    exitCode = 99;

    // parse args
    ok = parseArgs(argDesc, &argc, argv);
    if (!ok || argc < 2 || argc > 3 || printVersion || printHelp) {
        fprintf(stderr, "pdftops version %s\n", PACKAGE_VERSION);
        fprintf(stderr, "%s\n", popplerCopyright);
        fprintf(stderr, "%s\n", xpdfCopyright);
        if (!printVersion) {
            printUsage("pdftops", "<PDF-file> [<PS-file>]", argDesc);
        }
        if (printVersion || printHelp)
            exit(0);
        else
            exit(1);
    }
    if ((level1 ? 1 : 0) + (level1Sep ? 1 : 0) + (level2 ? 1 : 0) + (level2Sep ? 1 : 0) + (level3 ? 1 : 0) + (level3Sep ? 1 : 0) > 1) {
        fprintf(stderr, "Error: use only one of the 'level' options.\n");
        exit(1);
    }
    if ((doEPS ? 1 : 0) + (doForm ? 1 : 0) > 1) {
        fprintf(stderr, "Error: use only one of -eps, and -form\n");
        exit(1);
    }
    if (level1) {
        level = psLevel1;
    } else if (level1Sep) {
        level = psLevel1Sep;
    } else if (level2Sep) {
        level = psLevel2Sep;
    } else if (level3) {
        level = psLevel3;
    } else if (level3Sep) {
        level = psLevel3Sep;
    } else {
        level = psLevel2;
    }
    if (doForm && level < psLevel2) {
        fprintf(stderr, "Error: forms are only available with Level 2 output.\n");
        exit(1);
    }
    mode = doEPS ? psModeEPS : doForm ? psModeForm : psModePS;
    fileName = new GooString(argv[1]);

    // read config file
    globalParams = std::make_unique<GlobalParams>();
    if (origPageSizes) {
        paperWidth = paperHeight = -1;
    }
    if (paperSize[0]) {
        if (origPageSizes) {
            fprintf(stderr, "Error: -origpagesizes and -paper may not be used together.\n");
            exit(1);
        }
        if (!setPSPaperSize(paperSize, paperWidth, paperHeight)) {
            fprintf(stderr, "Invalid paper size\n");
            delete fileName;
            goto err0;
        }
    }
    if (overprint) {
        globalParams->setOverprintPreview(true);
    }
    if (expand) {
        globalParams->setPSExpandSmaller(true);
    }
    if (noShrink) {
        globalParams->setPSShrinkLarger(false);
    }
    if (level1 || level1Sep || level2 || level2Sep || level3 || level3Sep) {
        globalParams->setPSLevel(level);
    }
    if (quiet) {
        globalParams->setErrQuiet(quiet);
    }

#ifdef HAVE_SPLASH
    if (!processcolorformatname.toStr().empty()) {
        if (processcolorformatname.toStr() == "MONO8") {
            processcolorformat = splashModeMono8;
        } else if (processcolorformatname.toStr() == "CMYK8") {
            processcolorformat = splashModeCMYK8;
        } else if (processcolorformatname.toStr() == "RGB8") {
            processcolorformat = splashModeRGB8;
        } else {
            fprintf(stderr, "Error: Unknown process color format \"%s\".\n", processcolorformatname.c_str());
            goto err05;
        }
    }

#    ifdef USE_CMS
    if (!processcolorprofilename.toStr().empty()) {
        processcolorprofile = make_GfxLCMSProfilePtr(cmsOpenProfileFromFile(processcolorprofilename.c_str(), "r"));
        if (!processcolorprofile) {
            fprintf(stderr, "Error: Could not open the ICC profile \"%s\".\n", processcolorprofilename.c_str());
            goto err05;
        }
        if (!cmsIsIntentSupported(processcolorprofile.get(), INTENT_RELATIVE_COLORIMETRIC, LCMS_USED_AS_OUTPUT) && !cmsIsIntentSupported(processcolorprofile.get(), INTENT_ABSOLUTE_COLORIMETRIC, LCMS_USED_AS_OUTPUT)
            && !cmsIsIntentSupported(processcolorprofile.get(), INTENT_SATURATION, LCMS_USED_AS_OUTPUT) && !cmsIsIntentSupported(processcolorprofile.get(), INTENT_PERCEPTUAL, LCMS_USED_AS_OUTPUT)) {
            fprintf(stderr, "Error: ICC profile \"%s\" is not an output profile.\n", processcolorprofilename.c_str());
            goto err05;
        }
        displayprofilecolorspace = cmsGetColorSpace(processcolorprofile.get());
        if (displayprofilecolorspace == cmsSigCmykData) {
            if (processcolorformat == splashModeUndefined) {
                processcolorformat = splashModeCMYK8;
            } else if (processcolorformat != splashModeCMYK8) {
                fprintf(stderr, "Error: Supplied ICC profile \"%s\" is not a CMYK profile, but process color format is CMYK.\n", processcolorprofilename.c_str());
                goto err05;
            }
        } else if (displayprofilecolorspace == cmsSigGrayData) {
            if (processcolorformat == splashModeUndefined) {
                processcolorformat = splashModeMono8;
            } else if (processcolorformat != splashModeMono8) {
                fprintf(stderr, "Error: Supplied ICC profile \"%s\" is not a monochrome profile, but process color format is monochrome.\n", processcolorprofilename.c_str());
                goto err05;
            }
        } else if (displayprofilecolorspace == cmsSigRgbData) {
            if (processcolorformat == splashModeUndefined) {
                processcolorformat = splashModeRGB8;
            } else if (processcolorformat != splashModeRGB8) {
                fprintf(stderr, "Error: Supplied ICC profile \"%s\" is not a RGB profile, but process color format is RGB.\n", processcolorprofilename.c_str());
                goto err05;
            }
        }
    }
#    endif

    if (processcolorformat != splashModeUndefined) {
        if (level1 && processcolorformat != splashModeMono8) {
            fprintf(stderr, "Error: Setting -level1 requires -processcolorformat MONO8");
            goto err05;
        } else if ((level1Sep || level2Sep || level3Sep || overprint) && processcolorformat != splashModeCMYK8) {
            fprintf(stderr, "Error: Setting -level1sep/-level2sep/-level3sep/-overprint requires -processcolorformat CMYK8");
            goto err05;
        }
    }
#endif

    // open PDF file
    if (ownerPassword[0] != '\001') {
        ownerPW = new GooString(ownerPassword);
    } else {
        ownerPW = nullptr;
    }
    if (userPassword[0] != '\001') {
        userPW = new GooString(userPassword);
    } else {
        userPW = nullptr;
    }
    if (fileName->cmp("-") == 0) {
        delete fileName;
        fileName = new GooString("fd://0");
    }

    doc = PDFDocFactory().createPDFDoc(*fileName, ownerPW, userPW);

    if (userPW) {
        delete userPW;
    }
    if (ownerPW) {
        delete ownerPW;
    }
    if (!doc->isOk()) {
        exitCode = 1;
        goto err1;
    }

#ifdef ENFORCE_PERMISSIONS
    // check for print permission
    if (!doc->okToPrint()) {
        error(errNotAllowed, -1, "Printing this document is not allowed.");
        exitCode = 3;
        goto err1;
    }
#endif

    // construct PostScript file name
    if (argc == 3) {
        psFileName = new GooString(argv[2]);
    } else if (fileName->cmp("fd://0") == 0) {
        error(errCommandLine, -1, "You have to provide an output filename when reading from stdin.");
        goto err1;
    } else {
        const char *p = fileName->c_str() + fileName->getLength() - 4;
        if (!strcmp(p, ".pdf") || !strcmp(p, ".PDF")) {
            psFileName = new GooString(fileName->c_str(), fileName->getLength() - 4);
        } else {
            psFileName = fileName->copy();
        }
        psFileName->append(doEPS ? ".eps" : ".ps");
    }

    // get page range
    if (firstPage < 1) {
        firstPage = 1;
    }
    if (lastPage < 1 || lastPage > doc->getNumPages()) {
        lastPage = doc->getNumPages();
    }
    if (lastPage < firstPage) {
        error(errCommandLine, -1, "Wrong page range given: the first page ({0:d}) can not be after the last page ({1:d}).", firstPage, lastPage);
        goto err2;
    }

    // check for multi-page EPS or form
    if ((doEPS || doForm) && firstPage != lastPage) {
        error(errCommandLine, -1, "EPS and form files can only contain one page.");
        goto err2;
    }

    for (int i = firstPage; i <= lastPage; ++i) {
        pages.push_back(i);
    }

    // write PostScript file
    psOut = new PSOutputDev(psFileName->c_str(), doc, nullptr, pages, mode, paperWidth, paperHeight, noCrop, duplex);
    if (noCenter) {
        psOut->setPSCenter(false);
    }

    if (rasterAntialiasStr[0]) {
        if (!GlobalParams::parseYesNo2(rasterAntialiasStr, &rasterAntialias)) {
            fprintf(stderr, "Bad '-aaRaster' value on command line\n");
        }
    }

    if (forceRasterizeStr[0]) {
        PSForceRasterize forceRasterize = psRasterizeWhenNeeded;
        if (strcmp(forceRasterizeStr, "whenneeded") == 0) {
            forceRasterize = psRasterizeWhenNeeded;
        } else if (strcmp(forceRasterizeStr, "always") == 0) {
            forceRasterize = psAlwaysRasterize;
        } else if (strcmp(forceRasterizeStr, "never") == 0) {
            forceRasterize = psNeverRasterize;
        } else {
            fprintf(stderr, "Bad '-rasterize' value on command line\n");
        }
        psOut->setForceRasterize(forceRasterize);
    }

    if (splashResolution > 0) {
        psOut->setRasterResolution(splashResolution);
    }
#ifdef HAVE_SPLASH
    psOut->setProcessColorFormat(processcolorformat);
#    ifdef USE_CMS
    psOut->setDisplayProfile(processcolorprofile);
#    endif
#endif
    psOut->setEmbedType1(!noEmbedT1Fonts);
    psOut->setEmbedTrueType(!noEmbedTTFonts);
    psOut->setEmbedCIDPostScript(!noEmbedCIDPSFonts);
    psOut->setEmbedCIDTrueType(!noEmbedCIDTTFonts);
    psOut->setFontPassthrough(fontPassthrough);
    psOut->setPreloadImagesForms(preload);
    psOut->setOptimizeColorSpace(optimizeColorSpace);
    psOut->setPassLevel1CustomColor(passLevel1CustomColor);
#ifdef OPI_SUPPORT
    psOut->setGenerateOPI(doOPI);
#endif
    psOut->setUseBinary(psBinary);

    psOut->setRasterAntialias(rasterAntialias);
    if (psOut->isOk()) {
        for (int i = firstPage; i <= lastPage; ++i) {
            doc->displayPage(psOut, i, 72, 72, 0, noCrop, !noCrop, true);
        }
    } else {
        delete psOut;
        exitCode = 2;
        goto err2;
    }
    delete psOut;

    exitCode = 0;

    // clean up
err2:
    delete psFileName;
err1:
    delete doc;
err05:
    delete fileName;
err0:

    return exitCode;
}
